dao/exec.go

```diff
@@ -5,35 +5,59 @@
 package dao
 
 import (
+	"database/sql"
 	"fmt"
-
-	"github.com/openmarketplaceengine/openmarketplaceengine/cfg"
 )
 
-type RawSQL struct {
-	sql []string
+// Executor interface depicts ExecContext from
+// sql.DB, sql.Conn, sql.Tx, sql.Stmt.
+type Executor interface {
+	ExecContext(ctx Context, query string, args ...interface{}) (Result, error)
+}
+
+// Executable implementers perform actual Executor.ExecContext calls.
+type Executable interface {
+	Execute(ctx Context, exe Executor) error
 }
 
 //-----------------------------------------------------------------------------
 
-func (r *RawSQL) Appendf(format string, args ...interface{}) *RawSQL {
-	r.sql = append(r.sql, fmt.Sprintf(format, args...))
-	return r
+// SQLExec represents plain SQL statement without any parameter bindings.
+type SQLExec string
+
+func SQLExecf(format string, args ...interface{}) SQLExec {
+	return SQLExec(fmt.Sprintf(format, args...))
+}
+
+func (e SQLExec) Execute(ctx Context, exe Executor) error {
+	_, err := exe.ExecContext(ctx, string(e))
+	return err
 }
 
+func (e SQLExec) String() string {
+	return string(e)
+}
+
+//-----------------------------------------------------------------------------
+// Executable Runners
 //-----------------------------------------------------------------------------
 
-func (r *RawSQL) Exec(sql ...string) (err error) {
-	if failInit(&err) {
+// ExecDB runs executables with sql.DB.
+func ExecDB(ctx Context, execs ...Executable) error {
+	return WithConn(ctx, func(ctx Context, con *sql.Conn) (err error) {
```

Comment:
Will this call to `WithConn` repeatedly connect/disconnect to the database? Or is there a connection pool attached?

Reply:
No, it will not call WithConn repeatedly if used properly.
`exec ...Executable` is an open array parameters with each element called within one open connection.

`ExecDB(ctx, "INSERT", "UPDATE", "DELETE", ...)`

Almost always `ExecTX` should be used for auto transaction management. ExecDB is for something like `CREATE SCHEMA IF NOT EXISTS "ome"`